### DISCUSSION TITLE
Suport for Adafruit Feather RP2040

### DISCUSSION THREAD

**Discussion by fred-rabelo** [tttapa/Control-Surface#833] [Q&A]
Hello,

has anyone succeeded using an Adafruit Feather RP2040 board in any project?

Just trying to use the library i get to compile a simple file as the first example code:

#include <Control_Surface.h>
USBMIDI_Interface midi;
CCPotentiometer pot { A0, MIDI_CC::General_Purpose_Controller_1 };
void setup() { Control_Surface.begin(); }
void loop() { Control_Surface.loop(); }

but i receive the message from Windows:
"USB device not recognized - USB device malfunctioned, and Windows dos not recognize it"

I am guessing it has something to do with the MBED architecture, but i am a relatively noob and am a little lost here.
Since the Pi Pico is supported i thought the Adafruit version would also be possible to get running.

I am using the earlephilhower/arduino-pico core and  Adafruit TinyUSB library is enabled under "USB Stack".

I also tried to throw a look at [Custom-USB-MIDI-Backend page](https://tttapa.github.io/Control-Surface-doc/Doxygen/d7/d5f/Custom-USB-MIDI-Backend_8ino-example.html), but i don´t understand the page or the code and how i could use that to create a Custom Backend for the Adafruit Feather RP2040.

am I not getting something basic here? or is it really not that trivial?

Any help appreciated!
Thanks!

**Comment by tttapa** [comment]
Does the Adafruit TinyUSB example work correctly?  
https://github.com/adafruit/Adafruit_TinyUSB_Arduino/blob/5be0b21c0a5cf396b139bce0c6c0d403df051b0a/examples/MIDI/midi_test/midi_test.ino

**Reply by fred-rabelo** [reply]
Thanks for the reply.

the example on the link works correctly.

I tried after that to upload again the first example code from Control_Surface, but still receive the same error.

I noticed that when compiling i get a Warning on the Serial Monitor (i guess its the serial monitor):
WARNING: library Control Surface claims to run on avr, sam, samd, teensy, esp32, esp8266, megaavr, mbed, mbed_nano, mbed_rp2040 architecture(s) and may be incompatible with your current board which runs on rp2040 architecture(s).

**Reply by tttapa** [reply]
It seems to work on Linux, I don't have a Windows machine handy to test it at the moment.

You could try adding the following at the end of your `setup()`:
```cpp
  // wait until device mounted
  while(!TinyUSBDevice.mounted())
    delay(1);
```

Control Surface uses just a thin wrapper around the Adafruit TinyUSB library, so it shouldn't be too hard to get it to work:
https://tttapa.github.io/Control-Surface-doc/Doxygen/dc/dc1/USBMIDI__Adafruit__TinyUSB_8hpp_source.html

**Reply by fred-rabelo** [reply]
adding the code snippet only works adding also #include <Adafruit_TinyUSB.h>.

> #include <Control_Surface.h>
> #include <Adafruit_TinyUSB.h>
> 
> void setup() { Control_Surface.begin();
> // wait until device mounted
> while(!TinyUSBDevice.mounted())
> delay(1); }
> void loop() { Control_Surface.loop(); }

with it, windows recognizes the board as "USB Serial Device" (just as with Adafruit TinyUSB example) but the board is not recognized by the software as a midi device.
I am using the "Helm Synth" as a test software. With the Adafruit TinyUSB example the board appears directly as "Feather RP2040".

Is there some documentation i could read to help me understand how i could make the librarie work with this board?

Thanks so much for the help again!

**Reply by fred-rabelo** [reply]
I still haven't figured out what could be Generating the problem by the Adafruit feather rp2040 =\

Any ideas from what I could still try, or maybe tests I could run to get closer to the problem?

I started coding something with the adafruit_midi Circuitpython library. But I'm coding for pro tools, so I'm stuck with HUI protocol.
I could get the "handshake"  to work, the fader controlling pro tools and vice versa, but I'm still overwhelmed with the sysex messages...

Any help with the protocol would be also very much appreciated!!

**Comment by tttapa** [comment]
I've just pushed a fix that only accesses the `Adafruit_USBD_MIDI` instance if the USB device is mounted. This at least prevents crashes (unless you're trying to do a Serial.read()` before the Serial port is mounted, but that's a RP2040 issue for another time).

With that change, the following code works fine for me:
```cpp
#include <Control_Surface.h>

USBMIDI_Interface midi;

void setup() {
  Control_Surface.begin();
}

void loop() {
  static uint8_t counter = 0;
  static Timer<millis> timer {1000};
  if (timer)
    Control_Surface.sendControlChange(0x10, (counter++) & 0x7F);
  Control_Surface.loop();
}
```
Please verify that it also works on your end. It should send an increasing MIDI Control Change message every second.

If your own code still crashes, please post the full, exact code that causes the issue.

**Reply by fred-rabelo** [reply]
Thanks for the update!

Sadly the problem persists.

I did a "factory reset" on the board as per adafruit (flash_nuke_Factory-Reset.uf2) and downloaded and installed the Control Surface Lib again.

With the following code i get again the error message "USB device not recognized - USB device malfunctioned, and Windows dos not recognize it"

> #include <Control_Surface.h>
> 
> USBMIDI_Interface midi;
> 
> void setup() {
>   Control_Surface.begin();
> }
> 
> void loop() {
>   static uint8_t counter = 0;
>   static Timer<millis> timer {1000};
>   if (timer)
>     Control_Surface.sendControlChange(0x10, (counter++) & 0x7F);
>   Control_Surface.loop();
> }

as Output i get the following uploading the code above:

> WARNING: library Control Surface claims to run on avr, sam, samd, teensy, esp32, esp8266, megaavr, mbed, mbed_nano, mbed_rp2040 architecture(s) and may be incompatible with your current board which runs on rp2040 architecture(s).
> Sketch uses 62604 bytes (0%) of program storage space. Maximum is 8384512 bytes.
> Global variables use 11204 bytes (4%) of dynamic memory, leaving 250940 bytes for local variables. Maximum is 262144 bytes.
> Resetting COM5
> Converting to uf2, output size: 159232, start address: 0x2000
> Flashing D: (RPI-RP2)
> Wrote 159232 bytes to D:/NEW.UF2

and a few seconds after the error message from Windows.

I then did another factory reset and tried the following code as suggested before:

> #include <Control_Surface.h>
> #include <Adafruit_TinyUSB.h>
> 
> USBMIDI_Interface midi;
> 
> void setup() { Control_Surface.begin();
> // wait until device mounted
> while(!TinyUSBDevice.mounted())
> delay(1); }
> 
> void loop() {
>   static uint8_t counter = 0;
>   static Timer<millis> timer {1000};
>   if (timer)
>     Control_Surface.sendControlChange(0x10, (counter++) & 0x7F);
>   Control_Surface.loop();
> }

the error does not appear but i still cant see the board as a midi device, although the computer recognizes the COM port and i can write new code to the board with the Arduino IDE.

Could it be something with my Windows?

**Reply by tttapa** [reply]
> and downloaded and installed the Control Surface Lib again

What does your current `USBMIDI_Adafruit_TinyUSB.hpp` file look like?  
https://github.com/tttapa/Control-Surface/blob/main/src/MIDI_Interfaces/USBMIDI/USBMIDI_Adafruit_TinyUSB.hpp

**Reply by fred-rabelo** [reply]
like this:

```
#include <AH/Containers/Array.hpp>
#include <Settings/NamespaceSettings.hpp>

#include <Adafruit_TinyUSB.h>

BEGIN_CS_NAMESPACE

struct Adafruit_TinyUSB_USBDeviceMIDIBackend {
    using MIDIUSBPacket_t = AH::Array<uint8_t, 4>;
    void begin() {
#if defined(ARDUINO_ARCH_MBED) && defined(ARDUINO_ARCH_RP2040)
        // Manual begin() is required on core without built-in support for TinyUSB such as mbed rp2040
        TinyUSB_Device_Init(0);
#endif
        backend.begin();
    }
    MIDIUSBPacket_t read() {
        MIDIUSBPacket_t packet {};
        if (TinyUSBDevice.mounted())
            backend.readPacket(packet.data);
        return packet;
    }
    void write(MIDIUSBPacket_t packet) {
        if (TinyUSBDevice.mounted())
            backend.writePacket(packet.data);
    }
    void sendNow() {
        if (TinyUSBDevice.mounted())
            backend.flush();
    }
    bool preferImmediateSend() { return false; }

    Adafruit_USBD_MIDI backend;
};

END_CS_NAMESPACE
```

**Reply by fred-rabelo** [reply]
I found a work around!

had to install the official "Arduino Mbed" core in the place of the "earlephilhower-core", and no longer use "Adafruit_TinyUSB.h". (guess i´m using the standard Arduino USB library?)

The computer sees the board as Pi Pico instead of Feather Rp2040, but it seems to working until now with exception of the Adafruit Neo.Pixel library.

Would be nice to use a core that officially supports the Adafruit Feather Rp2040 board, but if it works, it works!
